frameworks/moveos-stdlib/sources/event_queue.move

```diff
@@ -0,0 +1,217 @@
+        let object_id = object::custom_object_id<String, EventQueue<E>>(name);
+        if (!object::exists_object(object_id)) {
+            let event_queue_obj = object::new_with_id(name, EventQueue<E> {
+                head_sequence_number: 0,
+                tail_sequence_number: 0,
+            });
+            let subscribers : vector<ObjectID> = vector::empty();
+            object::add_field(&mut event_queue_obj, SUBSCRIBERS_KEY, subscribers);
+            //We transfer the display object to the moveos_std
+            //And the caller do not need to care about the event queue object
+            object::transfer_extend(event_queue_obj, @moveos_std);
+        };
+        object::borrow_mut_object_extend<EventQueue<E>>(object_id)
+    }
+
+    fun borrow_subscribers<E>(event_queue_obj: &Object<EventQueue<E>>) : &vector<ObjectID>{
+        object::borrow_field(event_queue_obj, SUBSCRIBERS_KEY)
+    }
+
+    fun borrow_mut_subscribers<E>(event_queue_obj: &mut Object<EventQueue<E>>) : &mut vector<ObjectID>{
+        object::borrow_mut_field(event_queue_obj, SUBSCRIBERS_KEY)
+    }
+
+    #[private_generics(E)]
+    public fun emit<E : copy + drop + store>(name: String, event: E) {
+        let event_queue_obj = event_queue<E>(name);
+        let subscribers = borrow_subscribers(event_queue_obj);
+        //We only write the event to the event queue when there are subscribers
+        if(vector::is_empty(subscribers)){
+            return
+        };
+        let head_sequence_number = object::borrow(event_queue_obj).head_sequence_number;
+        let now = timestamp::now_milliseconds();
+        let on_chain_event = OnChainEvent {
+            event: event,
+            emit_time: now,
+        };
+        object::add_field(event_queue_obj, head_sequence_number, on_chain_event);
+        object::borrow_mut(event_queue_obj).head_sequence_number = head_sequence_number + 1;
+        remove_expired_events(event_queue_obj);    
+    }
+
+    public fun consume<E: copy + drop + store>(subscriber_obj: &mut Object<Subscriber<E>>) : Option<E> {
+        let subscriber = object::borrow_mut(subscriber_obj);
+        let subscriber_sequence_number = subscriber.sequence_number;
+        let event_queue_obj = event_queue<E>(subscriber.queue_name);
+        let head_sequence_number = object::borrow(event_queue_obj).head_sequence_number;
+        let tail_sequence_number = object::borrow(event_queue_obj).tail_sequence_number;
+        
+        if (subscriber_sequence_number >= head_sequence_number) {
+            return option::none()
+        };
+        //If the subscriber sequence number is less than the tail sequence number,
+        //It means the subscriber is too slow to consume the event
+        //We should update the subscriber sequence number to the tail sequence number
+        subscriber_sequence_number = if (tail_sequence_number > subscriber_sequence_number) {
+            tail_sequence_number
```

Comment:
I feel that it is not appropriate to jump directly to the end of the queue here, which may cause inconsistent data on the consumer side. It feels better to return an error.

Reply:
How does the subscriber handle the error?